Add new rule  `no-deprecated-router-transition-methods`

rtablada opened the pull request:
Similar to #1714 this implements a lint rule to detect and fix errors associated with the [`routing.transition-methods` deprecation](https://deprecations.emberjs.com/v3.x/#toc_routing-transition-methods).

This closes #1074

lin-ll commented on lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,190 @@
+    docs: {
+      description: 'enforce usage of router service transition methods',
+      category: 'Deprecations',
+      recommended: true,
```
Note: recommended rules are only added in major versions so this should be removed. We can enable it in the next major version.

lin-ll commented on lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,190 @@
+}
+
+const ERROR_MESSAGE =
+  'Do not rely on implicit service injections, these were deprecated in Ember 3.26 and will not work in 4.0.';
```
Can we make this error message more specific to the deprecation of implicit router transition methods?

rtablada replied:
I've updated the error message to create a more descriptive and matching message based on usage. The error before was a bad copy/paste from my other rule.

lin-ll commented on lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,193 @@
+  return { fixSteps, routerServicePropertyName };
+}
+
+const createErrorMessage = (methodUsed, desiredMethod, moduleType) =>
+  `Calling "${methodUsed}" in ${moduleType}s is deprecated, call ${desiredMethod} on the injected router service instead.`;
```
Instead of this, you can try using `messageId` instead (https://eslint.org/docs/latest/developer-guide/working-with-rules#messageids) and then you can pass the necessary dynamic data

rtablada replied:
Cool, I'm not too familiar with custom messages, this will be helpful and I aim to fix this today.

lin-ll commented on tests/lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,321 @@
+//------------------------------------------------------------------------------
```
Do you have any test cases for multiple classes in one file? Maybe some with correct usage and some with incorrect usage?

Also, have you tried testing this on a real codebase?

rtablada replied:
I've tested this on our codebase of over 300 components and 200+ routes and it was able to identify and autofix all instances (109 usecases).

I'll add some test cases for nested and multiple classes to confirm proper identification and fixes for those.

bmish commented on docs/rules/no-deprecated-router-transition-methods.md:
```diff
@@ -0,0 +1,96 @@
+
+## Related Rules
+
+- [no-unnecessary-service-injection-argument](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-unnecessary-service-injection-argument.md) omit service injection argument if possible
+- [no-implicit-service-injection-argument](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-unnecessary-service-injection-argument.md) require the service injection argument for all services (fix output for this rule follows this pattern)
```
On second thought, these rules don't really seem related? It would be nice if we handled either style since these are just stylistic preferences.

rtablada replied:
Not sure if it's too much of a round trip, but from my testing autofix will update appropriately either way. I'm not sure if we need to mention it in this case here at all then?

rtablada commented:
I've updated the tests to have some examples of multiple classes (and fixes), updated the readme as requested, and have made changes to the error message to use `messageId`

lin-ll commented:
LGTM - but looks like some tests / lint errors are still failing

bmish commented on docs/rules/no-deprecated-router-transition-methods.md:
```diff
@@ -0,0 +1,91 @@
+
+## Migration
+
+The autofixer for this rule will update classes and add injections for the configured services.
```
Perhaps omit "configured" since this rule doesn't have any configuration?

bmish commented on tests/lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,435 @@
+      import Route from '@ember/routing/route';
+      import { inject as service } from '@ember/service';
+      export default class SettingsRoute extends Route {
+        @service() router;
```
Let's also add a test that `@service router;` is detected properly?

rtablada replied:
Yup similar to implicit. I'm going to add checks for all the different types of service injection as well as legacy `.extends` which I know currently are not covered in this PR.

bmish commented on tests/lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,435 @@
+});
+
+ruleTester.run('no-deprecated-router-transition-methods', rule, {
+  valid: [
```
It looks like your rule is supposed to ignore code outside an Ember class, can we test that?

bmish commented on lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,201 @@
+    const classStack = new Stack();
+
+    // This rule does not apply to test files or non class modules
+    if (emberUtils.isTestFile(context.getFilename())) {
```
According to test coverage checks, this line is not tested. Can you test it? And also be sure to monitor the test coverage to ensure every line in the new rule is covered.

```
yarn test --coverage lib/rules/no-deprecated-router-transition-methods.js
```

rtablada replied:
After looking through, I'm going to remove this test check.

There are cases where you may have module definitions in tests and we don't want to ignore them.

Though it's unlikely for teams to define mock controllers/routes, it is still possible.

bmish commented on lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,201 @@
+      },
+
+      MemberExpression(node) {
+        if (!isValidModule) {
```
According to test coverage checks, this line is not tested. Can you test it? And also be sure to monitor the test coverage to ensure every line in the new rule is covered.

```
yarn test --coverage lib/rules/no-deprecated-router-transition-methods.js
```

bmish commented on docs/rules/no-deprecated-router-transition-methods.md:
```diff
@@ -0,0 +1,91 @@
+
+The autofixer for this rule will update classes and add injections for the configured services.
+
+## References
```
Can you add links to the relevant router function specs?

bmish commented on tests/lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,435 @@
+        },
+      ],
+    },
+    {
```
Since these test cases are pretty bulky, a brief comment in front of each could be helpful to indicate the goal of each test case (what it's trying to test).

bmish reviewed (changes requested):
Great progress. Left comments mostly about tests.

bmish commented on tests/lib/rules/no-deprecated-router-transition-methods.js:
```diff
@@ -0,0 +1,435 @@
+  },
+});
+
+ruleTester.run('no-deprecated-router-transition-methods', rule, {
```
Since we're using the class stack, technically we should also be testing that the behavior is correct with nested classes. For example, that the behavior is correct inside the nested class, and after the nested class but still inside the outer class.

rtablada replied:
Similar to the other PR, I'll add nested classes and sibling classes.

rtablada commented:
Updated this PR with requested feedback and learnings from the `no-implicit-injections` PR.

Added nested class tests, class expression tests, and accounted for legacy `.extends`

bmish reviewed (approved):
Thank you!
